### ISSUE TITLE
Execute flow  jobs in sequence order.

### ISSUE THREAD

**Issue by nordicgit70** [taskforcesh/bullmq#1858]
Currently jobs can be executed in sequence order by using the [steps](https://docs.bullmq.io/patterns/process-step-jobs) pattern. In my use case I typically break-down jobs into smaller steps to reduce complexity and have atomic units-of-work with a specific responsibility.

As a solution I would like to be able to set an `order` value that determines the order in which jobs in the flow are executed. By default the order can be zero (no dependencies). A mechanism can be to check the state of all children jobs and execute the jobs with the lowest order number. If a specific job fails the next jobs are not executed.

The alternative is to add jobs as nested children with the first job in the sequence as nested deepest child.  This will cause a roll-up execution of all jobs to the final parent.

Motivation would be that a sequential execution is a common use-case and add more value to the flow concept.

**Comment by nordicgit70** [comment]
My current workaround to execute jobs in sequence is a nested flow (alternative solution). 
The worker executes different steps based on the assigned name.

```
export function createSerialJob(obj): FlowJob {
    return {
        name: JobNames.Parent,
        data: obj,
        queueName: myQueue.name,
        children: [
            {
                name: JobNames.Step2,
                data: obj,
                queueName: myQueue.name,
                children: [
                    {
                        name: JobNames.Step1,
                        data: obj,
                        queueName: myQueue.name,
                    },
                ],
            },
        ],
        opts: { jobId: obj.Id },
    };
}
```

**Comment by manast** [comment]
This could be solved with a helper function that takes an array with the sequence and generates the nested structure. At the end of the day, internally, this is how it must work anyway, as there is no concept of order in BullMQ other than the position in the queue.

**Comment by HarikalarKutusu** [comment]
I had to (halfway) implement a rudimentary orchestrator in the last couple of days to handle multi-service / multi-possibility job trees to handle [my problem](https://github.com/taskforcesh/bullmq/issues/1855). It actually does include that helper function @manast mentioned. Some construct ideas:
```
// Following strings should be in format serviceName::jobName for simplicity
export interface IOrcestratorNode {
  series?: string[]; // to be IOrcestratorNode[]
  parallel?: string[]; // to be IOrcestratorNode[]
}

export interface IOrcestrator {
  finished: string;
  single?: string;
  flow?: IOrcestratorNode[];
}
```
At the end of a successful job run, I call the "helper" function `maybeChain(from)` which forms the tree and pushes it to job queues. I handle the data flow through redis and/or disk files. This way I can:
````
from => single
from => parallel(list)
from => series(list)
from => parallel(
                  single
                  series(list1)
                  series(list2)
             )
...etc...
```
It is not at all dynamic, but it was the only way I could think of to centralize a flow like this one:

![image](https://user-images.githubusercontent.com/8849617/235347224-d0933857-b54d-400b-b658-c6109907c029.png)

**Comment by zwhitchcox** [comment]
Nice, this is my solution to automatically make the steps in order:

```ts
const steps = [
  CallStep.Recording,
  CallStep.Transcription
]

const createCallFlow = (rootId: string, callId: string) => {
  const root: any = {};
  let cur = root;
  for (const step of steps) {
    cur.name = step;
    cur.queueName = name;
    cur.data = {
      rootId,
      callId,
    };
    if (step !== steps[steps.length - 1]) {
      cur.children = [cur = {}]
    }
  }
  return root.children[0];
}

export const flow = async (options: any) => {
  const calls = await getCalls(options.since, options.until);
  const rootId = Math.random().toString(36).slice(2);
  const queueName = name;
  return {
    queueName,
    name: 'end',
    children: calls.map(({ id:callId }) => createCallFlow(rootId, String(callId))),
  };
}
```

Basically the helper function that @manast mention...this could be easily altered to support parallel processing by adding multiple children in the array, but the problem is there is no way to represent multiple steps depending on a single step. That is, a single step can have multiple children, but a child can only have one parent...this is a little limiting. Right now, it's fine for me to just process one step at a time, even for jobs that could be done in parallel, but orchestration functionality like @HarikalarKutusu mentions would be nice. The only way I can see to do that though is to have a listener on each queue that checks to see if it has any more jobs to start or have multiple steps being processed by a single worker (which is not ideal, because you don't get the parallel compute power of multiple workers.

Not sure if there's a better way to do this.

**Comment by HarikalarKutusu** [comment]
> but a child can only have one parent

Right... A "join" node (wait-for) is required, using a listener or observable method, or just by polling if it is not time-sensitive.
This might be implemented using Redis to keep flags and/or counters with a locking mechanism.

In my simple stuff presented above, assume there are tasks A, B, C, and C should wait for A and B to be completed.
- When A finished it sets A-Finished flag and calls C, C checks if A & B are both finished, as it is not, so it exits
- When B finished it sets B-Finished flag and calls C, C checks if A & B are both finished, both are finished, it resets these flags and runs

**Comment by CodingMeSwiftly** [comment]
> but a child can only have one parent

My two cents:
I found that you can actually have two parents for a job, if you define the jobIds yourself. You can create multiple 'subtrees' of child jobs simultaneously. If two or more of those subtrees have a child with matching `opts.jobId`, BullMQ will correctly push only one instance of that job to the queue (a queue can't have two jobs with the same jobId) and install dependencies for both parents. When the child job finishes, both parents will run. I haven't tested the behaviour for when the subtrees link up upstream, but since dependencies are just a linked list, there should be no problem with that.
